partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/ExcludeUtils.kt

```diff
@@ -0,0 +1,159 @@
+package org.partiql.planner.internal.typer
+
+import org.partiql.errors.Problem
+import org.partiql.errors.ProblemCallback
+import org.partiql.errors.UNKNOWN_PROBLEM_LOCATION
+import org.partiql.planner.PlanningProblemDetails
+import org.partiql.planner.internal.ir.Identifier
+import org.partiql.planner.internal.ir.Identifier.CaseSensitivity
+import org.partiql.planner.internal.ir.Rel
+import org.partiql.planner.internal.ir.Rex
+import org.partiql.types.AnyOfType
+import org.partiql.types.CollectionType
+import org.partiql.types.StaticType
+import org.partiql.types.StructType
+
+/**
+ * Checks for every exclude path in [paths], that a value in the [bindings] is excluded. If there is no value excluded,
+ * a warning is added to the [onProblem] callback.
+ */
+internal fun checkForInvalidExcludePaths(bindings: List<Rel.Binding>, paths: List<Rel.Op.Exclude.Item>, onProblem: ProblemCallback) {
+    paths.forEach { excludePath ->
+        val root = excludePath.root
+        val steps = excludePath.steps
+        val excludesSomething = bindings.any { binding ->
+            when (root) {
+                is Rex.Op.Var.Unresolved -> {
+                    when (val id = root.identifier) {
+                        is Identifier.Symbol -> {
+                            if (id.isEquivalentTo(binding.name)) {
+                                binding.type.checkExclude(steps)
+                            } else {
+                                false
+                            }
+                        }
+                        is Identifier.Qualified -> {
+                            if (id.root.isEquivalentTo(binding.name)) {
+                                binding.type.checkExclude(steps)
+                            } else {
+                                false
+                            }
+                        }
+                    }
+                }
+                is Rex.Op.Var.Resolved -> false
```

Comment:
Why do we return false here?

Reply:
When this check is run in `PlanTyper`, the root for each exclude path should be an unresolved var, so this case should not occur. The value returned here shouldn't matter.